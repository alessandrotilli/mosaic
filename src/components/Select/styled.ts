import { ListSubheader as MUIListSubheader, styled, TextField as MUITextField } from "@material-ui/core";

export const StyledMUIListSubheader = styled(MUIListSubheader)(({ theme }) => ({
  backgroundColor: theme.palette.background.default,
}));

export const StyledMUITextField = styled(MUITextField)({
  width: "100%",
});
