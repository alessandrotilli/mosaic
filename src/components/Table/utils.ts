export const COLUMN_CHECKBOX_PATH = "checkbox-selection";
export const COLUMN_ROW_ACTIONS_PATH = "row-actions";

export const TOOLBAR_HEIGHT = 64;
export const TOOLBAR_HEIGHT_MOBILE = 56;

export const PAGINATION_TOOLBAR_HEIGHT = 64;
