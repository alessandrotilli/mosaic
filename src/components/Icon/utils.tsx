import React from "react";
import MUIAccountCircleIcon from "@material-ui/icons/AccountCircle";
import MUIAddIcon from "@material-ui/icons/Add";
import MUIAppsIcon from "@material-ui/icons/Apps";
import MUIArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import MUIBlockIcon from "@material-ui/icons/Block";
import MUIBrightness4Icon from "@material-ui/icons/Brightness4";
import MUIBusinessIcon from "@material-ui/icons/Business";
import MUICheckIcon from "@material-ui/icons/Check";
import MUICheckBoxIcon from "@material-ui/icons/CheckBox";
import MUICheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import MUIChevronLeftIcon from "@material-ui/icons/ChevronLeft";
import MUIChevronRightIcon from "@material-ui/icons/ChevronRight";
import MUICloseIcon from "@material-ui/icons/Close";
import MUICropIcon from "@material-ui/icons/Crop";
import MUIDeleteIcon from "@material-ui/icons/Delete";
import MUIDetailsIcon from "@material-ui/icons/Details";
import MUIDraftsIcon from "@material-ui/icons/Drafts";
import MUIEditIcon from "@material-ui/icons/Edit";
import MUIErrorIcon from "@material-ui/icons/Error";
import MUIFilterListIcon from "@material-ui/icons/FilterList";
import MUIFirstPageIcon from "@material-ui/icons/FirstPage";
import MUIGroupWorkIcon from "@material-ui/icons/GroupWork";
import MUIHomeIcon from "@material-ui/icons/Home";
import MUIKeyboardArrowDownIcon from "@material-ui/icons/KeyboardArrowDown";
import MUIKeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp";
import MUILanguageIcon from "@material-ui/icons/Language";
import MUILastPageIcon from "@material-ui/icons/LastPage";
import MUIListIcon from "@material-ui/icons/List";
import MUIMailIcon from "@material-ui/icons/Mail";
import MUIMenuIcon from "@material-ui/icons/Menu";
import MUINotificationsIcon from "@material-ui/icons/Notifications";
import MUIOpenInNew from "@material-ui/icons/OpenInNew";
import MUIPauseIcon from "@material-ui/icons/Pause";
import MUIPlayArrowIcon from "@material-ui/icons/PlayArrow";
import MUIPlaylistAddCheckIcon from "@material-ui/icons/PlaylistAddCheck";
import MUIPowerSettingsNewIcon from "@material-ui/icons/PowerSettingsNew";
import MUIPrintIcon from "@material-ui/icons/Print";
import MUIRefreshIcon from "@material-ui/icons/Refresh";
import MUISaveIcon from "@material-ui/icons/Save";
import MUIScheduleIcon from "@material-ui/icons/Schedule";
import MUISearchIcon from "@material-ui/icons/Search";
import MUISendIcon from "@material-ui/icons/Send";
import MUISettingsIcon from "@material-ui/icons/Settings";
import MUISignalWifi1BarOutlinedIcon from "@material-ui/icons/SignalWifi1BarOutlined";
import MUITimelineIcon from "@material-ui/icons/Timeline";
import MUIVisibilityIcon from "@material-ui/icons/Visibility";
import MUIVisibilityOffIcon from "@material-ui/icons/VisibilityOff";
import MUIWarningIcon from "@material-ui/icons/Warning";
import MUIWifiOffOutlinedIcon from "@material-ui/icons/WifiOffOutlined";
import MUIWifiOutlinedIcon from "@material-ui/icons/WifiOutlined";

import { Icons, IIconsCatalog } from "../../types/Icon";

export const iconsCatalog: IIconsCatalog = {
  [Icons.account]: <MUIAccountCircleIcon />,
  [Icons.add]: <MUIAddIcon />,
  [Icons.apps]: <MUIAppsIcon />,
  [Icons.block]: <MUIBlockIcon />,
  [Icons.brightness]: <MUIBrightness4Icon />,
  [Icons.business]: <MUIBusinessIcon />,
  [Icons.chart_timeline]: <MUITimelineIcon />,
  [Icons.check]: <MUICheckIcon />,
  [Icons.checkbox]: <MUICheckBoxIcon />,
  [Icons.checkbox_empty]: <MUICheckBoxOutlineBlankIcon />,
  [Icons.clock]: <MUIScheduleIcon />,
  [Icons.close]: <MUICloseIcon />,
  [Icons.connection_slow]: <MUISignalWifi1BarOutlinedIcon />,
  [Icons.crop]: <MUICropIcon />,
  [Icons.delete]: <MUIDeleteIcon />,
  [Icons.details]: <MUIDetailsIcon />,
  [Icons.down]: <MUIKeyboardArrowDownIcon />,
  [Icons.edit]: <MUIEditIcon />,
  [Icons.error]: <MUIErrorIcon />,
  [Icons.filter]: <MUIFilterListIcon />,
  [Icons.first]: <MUIFirstPageIcon />,
  [Icons.group]: <MUIGroupWorkIcon />,
  [Icons.home]: <MUIHomeIcon />,
  [Icons.language]: <MUILanguageIcon />,
  [Icons.last]: <MUILastPageIcon />,
  [Icons.left]: <MUIChevronLeftIcon />,
  [Icons.list_check]: <MUIPlaylistAddCheckIcon />,
  [Icons.list]: <MUIListIcon />,
  [Icons.mail_open]: <MUIDraftsIcon />,
  [Icons.mail]: <MUIMailIcon />,
  [Icons.menu]: <MUIMenuIcon />,
  [Icons.notifications]: <MUINotificationsIcon />,
  [Icons.open_new]: <MUIOpenInNew />,
  [Icons.pause]: <MUIPauseIcon />,
  [Icons.play]: <MUIPlayArrowIcon />,
  [Icons.power]: <MUIPowerSettingsNewIcon />,
  [Icons.print]: <MUIPrintIcon />,
  [Icons.refresh]: <MUIRefreshIcon />,
  [Icons.right]: <MUIChevronRightIcon />,
  [Icons.save]: <MUISaveIcon />,
  [Icons.search]: <MUISearchIcon />,
  [Icons.send]: <MUISendIcon />,
  [Icons.settings]: <MUISettingsIcon />,
  [Icons.up_arrow]: <MUIArrowUpwardIcon />,
  [Icons.up]: <MUIKeyboardArrowUpIcon />,
  [Icons.visibility]: <MUIVisibilityIcon />,
  [Icons.visibility_off]: <MUIVisibilityOffIcon />,
  [Icons.warn]: <MUIWarningIcon />,
  [Icons.wifi_off]: <MUIWifiOffOutlinedIcon />,
  [Icons.wifi_on]: <MUIWifiOutlinedIcon />,
};
