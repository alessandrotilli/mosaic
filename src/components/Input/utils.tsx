import React from "react";
import { InputAdornment as MUIInputAdornment } from "@material-ui/core";

import { IInputAdornment } from "../../types/Input";
import IconButton from "../IconButton";

export const getAdornment = (adornment?: IInputAdornment) => {
  if (!adornment) {
    return;
  }

  const { icon, onClick } = adornment;
  return (
    <MUIInputAdornment position="end">
      <IconButton disabled={!onClick} icon={icon} onClick={onClick!} size="small" />
    </MUIInputAdornment>
  );
};
