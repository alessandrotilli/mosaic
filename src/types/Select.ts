import { ReactNode } from "react";

import { ILoadable, ILocalizable } from "./Base";
import { IInputField } from "./Input";

export interface SelectPopperProps {
  /**
   * Width of options list popup
   */
  popperWidth?: number;
}

interface SelectBaseProps<T> extends SelectPopperProps, ILocalizable, ILoadable, IInputField {
  /**
   * Enables autocomplete feature
   */
  autoComplete?: boolean;
  /**
   * Automatically sorts options in alphabetical order (localeCompare)
   */
  autoSort?: boolean;
  /**
   * Method to allow custom rendering of each option
   */
  customOptionRendering?: (option: T, selected: boolean) => ReactNode;
  /**
   * Method to determine group label from group name
   */
  getGroupLabel?: (groupName: string) => string;
  /**
   * Method to extract option label from option
   */
  getOptionLabel?: (option: T) => string;
  /**
   * Method to determine currently selected option
   */
  getOptionSelected?: (option: T, value: T) => boolean;
  /**
   * Method to group options using one option property
   */
  groupBy?: (option: T) => string;
  /**
   * Callback for close event
   */
  onClose?: () => void;
  /**
   * Callback for change event on input field
   */
  onInputChange?: (input: string) => void;
  /**
   * Callback for scrollEnd event on options list
   */
  onScrollEnd?: () => void;
  /**
   * List of available options
   */
  options: T[];
  /**
   * Number of options
   */
  optionsNumber?: number;
}

type SingleSelectDataType<T> = T | null;

interface SelectSingleProps<T> extends SelectBaseProps<T> {
  /**
   * Enables multiple mode
   */
  multiple: false;
  onChange: (value: SingleSelectDataType<T>) => void;
  value?: SingleSelectDataType<T>;
}

type MultipleSelectDataType<T> = T[] | null;

interface SelectMultipleProps<T> extends SelectBaseProps<T> {
  /**
   * Enables multiple mode
   */
  multiple: true;
  onChange: (value: MultipleSelectDataType<T>) => void;
  value?: MultipleSelectDataType<T>;
}

export type SelectProps<T> = SelectSingleProps<T> | SelectMultipleProps<T>;
