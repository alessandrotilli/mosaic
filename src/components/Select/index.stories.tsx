import React from "react";
import { action } from "@storybook/addon-actions";
import { boolean, text, number, select } from "@storybook/addon-knobs";
import Select, { SelectWithProps, DATA_CY_DEFAULT, DATA_CY_SHORTCUT, LOCALIZABLE_PROPS } from ".";
import { InputSize, InputVariant } from "../../types/Input";
import { getDocumentationPage } from "../../utils/stories";
import IntlProviderMock, { LocaleMock, MessageMock } from "../../utils/mocks/IntlProviderMock";

export default {
  title: "Select",
  component: SelectWithProps,
  parameters: {
    ...getDocumentationPage({
      basedOn: "@material-ui/core/Autocomplete",
      component: "Select",
      e2eTestInfo: {
        dataCyDefault: DATA_CY_DEFAULT,
        dataCyShortcut: DATA_CY_SHORTCUT,
      },
      localizableProps: LOCALIZABLE_PROPS,
    }),
  },
};

export const Canvas = () => (
  <Select
    autoComplete={boolean("autoComplete", true)}
    disabled={boolean("disabled", false)}
    label={text("label", "Label")}
    loading={boolean("loading", false)}
    multiple={boolean("multiple", false)}
    onChange={action("On Change")}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
    popperWidth={number("popperWidth", 300)}
  />
);

export const Basic = () => (
  <Select
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const Disabled = () => (
  <Select
    disabled
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const Multiple = () => (
  <Select
    label="Arts & Creativity"
    multiple
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const Grouped = () => (
  <Select
    groupBy={(option) => option.slice(0, 1)}
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const InitialValue = () => (
  <Select
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
    value="Mosaic"
  />
);

export const Loading = () => (
  <Select
    label="Arts & Creativity"
    loading
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const Localized = () => (
  // IntlProviderMock simulates external IntlProvider context
  <IntlProviderMock locale={select("locale", LocaleMock, LocaleMock.en)}>
    <Select
      label={MessageMock.title}
      localized
      multiple={false}
      onChange={(value) => {}}
      options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
      placeholder={MessageMock.title}
    />
  </IntlProviderMock>
);

export const Required = () => (
  <Select
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
    required
  />
);

export const Small = () => (
  <Select
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
    size={InputSize.small}
  />
);

export const Variant = () => (
  <Select
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
    variant={InputVariant.filled}
  />
);

export const WithCustomGroupLabel = () => (
  <Select
    getGroupLabel={(groupName) => `Letter: ${groupName}`}
    groupBy={(option) => option.slice(0, 1)}
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);

export const WithCustomOptionRendering = () => (
  <Select
    customOptionRendering={(option) => <b>{option.slice(0, 3).toUpperCase()}</b>}
    label="Arts & Creativity"
    multiple={false}
    onChange={(value) => {}}
    options={["Mosaic", "Murales", "Paintings", "Photography", "Sculpture"]}
  />
);
