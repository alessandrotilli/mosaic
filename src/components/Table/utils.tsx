import React, { forwardRef, Fragment } from "react";
import { Action as MTActionType, Column as MTColumnType, Options as MTOptionsType } from "material-table";
import { Icons, IconSize } from "../../types/Icon";
import { TableActionPosition, ITableAction, ITableColumn } from "../../types/Table";
import Button from "../Button";
import Icon from "../Icon";
import IconButton from "../IconButton";
import { getDataCyForSubComponent, suppressEvent } from "../../utils";

// For default props refer to Options section here https://material-table.com/#/docs/all-props
export const DEFAULT_TABLE_OPTIONS: MTOptionsType = {
  debounceInterval: 500,
  draggable: false,
  emptyRowsWhenPaging: false,
};

export const actionAdapter = (action: ITableAction): MTActionType<object> => {
  const {
    callback,
    label,
    disabled = false,
    hidden = false,
    icon = undefined,
    position = TableActionPosition.default,
  } = action;
  return {
    disabled,
    hidden,
    icon: `${label}$${icon}`,
    onClick: (event, data) => {
      if (event) {
        suppressEvent(event);
      }
      callback(data);
    },
    position,
  };
};

export const actionComponentAdapter = (props: any, dataCy: string) => {
  const { data, action } = props;
  const { disabled, hidden, icon, position, onClick }: MTActionType<object> = action;
  if (hidden) {
    return null;
  }

  const displaysButton = position === TableActionPosition.default;
  const labelAndIconName = icon as string;
  const label = labelAndIconName.split("$")[0];
  const iconName = labelAndIconName.split("$")[1] as Icons;
  return (
    <Fragment>
      {displaysButton && (
        <Fragment>
          <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
          <Button
            dataCy={getDataCyForSubComponent(dataCy, `action-${label}`)}
            disabled={disabled}
            icon={!iconName ? undefined : { name: iconName }}
            label={label}
            onClick={() => onClick(undefined, data)}
          />
        </Fragment>
      )}
      {!displaysButton && (
        <Fragment>
          <IconButton
            dataCy={getDataCyForSubComponent(dataCy, `action-${label}`)}
            disabled={disabled}
            icon={iconName}
            onClick={() => onClick(undefined, data)}
            size={IconSize.small}
          />
          <span>&nbsp;&nbsp;</span>
        </Fragment>
      )}
    </Fragment>
  );
};

export const columnAdapter = (column: ITableColumn): MTColumnType<object> => {
  const { label, path, render, width } = column;
  return {
    ...(!!width && ({ width } as object)),
    field: path,
    render,
    title: label,
  };
};

export const iconAdapter = (name: Icons, size = IconSize.default) =>
  forwardRef((props: any, ref: any) => {
    return <Icon forwarded={{ ...props, ref }} name={name} size={size} />;
  });
