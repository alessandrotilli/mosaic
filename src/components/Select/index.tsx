import React, { CSSProperties, Fragment, SyntheticEvent, useCallback, useMemo } from "react";
import {
  ListSubheader as MUIListSubheader,
  Popper as MUIPopper,
  PopperProps as MUIPopperProps,
  TextField as MUITextField,
  useTheme,
} from "@material-ui/core";
import { Autocomplete as MUIAutocomplete, Skeleton as MUISkeleton } from "@material-ui/lab";

import { InputSize, InputType, InputVariant } from "../../types/Input";
import { ISelect } from "../../types/Select";
import { getComposedDataCy, suppressEvent } from "../../utils";
import localized, { ILocalizableProperty } from "../../utils/hocs/localized";
import Checkbox from "../Checkbox";
import Typography from "../Typography";

export const DATA_CY_DEFAULT = "select";
export const DATA_CY_SHORTCUT = "label";
export const LOCALIZABLE_PROPS: ILocalizableProperty[] = [
  { name: "label", type: "string" },
  { name: "placeholder", type: "string" },
];

export const SUBPARTS_MAP = {
  loading: {
    label: "Loading",
  },
  optionCheckbox: {
    label: "Option Checkbox (with label)",
    value: (label = "{label}") => `option-${label}-checkbox`,
  },
  optionLabel: {
    label: "Option Label (with label)",
    value: (label = "{label}") => `option-${label}-label`,
  },
  optionGroupLabel: {
    label: "Option Group (with label)",
    value: (label = "{label}") => `option-group-${label}`,
  },
  outerWrapper: {
    label: "Outer Wrapper",
  },
};

const Select = <T extends any>({
  autoComplete = true,
  autoSort = false,
  customOptionRendering,
  dataCy = DATA_CY_DEFAULT,
  disabled = false,
  getGroupLabel,
  getOptionLabel: externalGetOptionLabel,
  getOptionSelected,
  groupBy,
  label,
  loading = false,
  multiple = false,
  onChange: externalOnChange,
  onInputChange: externalOnInputChange,
  onScrollEnd,
  options: externalOptions,
  placeholder,
  popperWidth,
  required = false,
  size = InputSize.default,
  style,
  type = InputType.default,
  value = null,
  variant = InputVariant.default,
}: ISelect<T>) => {
  const theme = useTheme();

  const getOptionLabel = useCallback(
    (option: T): string => (externalGetOptionLabel ? externalGetOptionLabel(option) : `${option}`),
    [externalGetOptionLabel]
  );

  const isOptionSelected = useCallback(
    (option: T, value: T): boolean => {
      if (getOptionSelected) {
        return getOptionSelected(option, value);
      }

      return !!value && option === value;
    },
    [getOptionSelected]
  );

  const onChange = useCallback(
    (event, value: any) => {
      suppressEvent(event);
      externalOnChange(value);
    },
    [externalOnChange]
  );

  const onInputChange = useCallback(
    (event, value: any) => {
      if (!externalOnInputChange) {
        return;
      }

      suppressEvent(event);
      externalOnInputChange!(value);
    },
    [externalOnInputChange]
  );

  const onScroll = useCallback(
    ({ currentTarget: listboxNode }: SyntheticEvent) => {
      if (!listboxNode || !onScrollEnd) {
        return;
      }

      const { clientHeight, scrollHeight, scrollTop } = listboxNode;
      const scrollEnded = clientHeight + scrollTop === scrollHeight;
      if (scrollEnded) {
        onScrollEnd();
      }
    },
    [onScrollEnd]
  );

  const options = useMemo(() => {
    let options = [...externalOptions];
    if (!autoSort && !groupBy) {
      return options;
    }

    if (groupBy) {
      options = options.sort((one: T, another: T) => {
        const oneGroup = groupBy(one);
        const anotherGroup = groupBy(another);
        return oneGroup.localeCompare(anotherGroup);
      });
    }

    if (autoSort) {
      options = options.sort((one: T, another: T) => {
        const oneLabel = getOptionLabel(one);
        const anotherLabel = getOptionLabel(another);
        return oneLabel.localeCompare(anotherLabel);
      });
    }

    return options;
  }, [autoSort, externalOptions, getOptionLabel, groupBy]);

  const isOptionSelectable = useCallback(
    (value: T | null): boolean => !!value && options.some((option) => isOptionSelected(option, value)),
    [options, isOptionSelected]
  );

  const validateValue = useCallback(
    (value: T | T[] | null): T | T[] | null => {
      if (multiple) {
        const isValidMultipleValue = Array.isArray(value) && value.length > 0 && value.every(isOptionSelectable);
        return isValidMultipleValue ? value : [];
      }

      const isValidValue = !Array.isArray(value) && isOptionSelectable(value);
      return isValidValue ? value : null;
    },
    [isOptionSelectable, multiple]
  );

  return (
    <MUIAutocomplete<T, boolean>
      autoComplete={autoComplete}
      data-cy={getComposedDataCy(dataCy, SUBPARTS_MAP.outerWrapper)}
      disableCloseOnSelect={multiple}
      disabled={disabled}
      getOptionLabel={getOptionLabel}
      getOptionSelected={isOptionSelected}
      groupBy={groupBy}
      ListboxProps={{
        onScroll,
        style: {
          padding: 0,
          width: "100%",
        },
      }}
      loading={loading}
      multiple={multiple}
      onChange={onChange}
      onInputChange={onInputChange}
      options={options}
      PopperComponent={(props: MUIPopperProps) => {
        const { anchorEl } = props;
        const anchorElRef = anchorEl as any;
        const anchorElWidth = anchorElRef ? anchorElRef.clientWidth : null;
        const width = !!popperWidth && popperWidth > anchorElWidth ? popperWidth : anchorElWidth;
        return <MUIPopper {...props} placement="bottom-start" style={{ width }} />;
      }}
      renderGroup={(groupProps) => {
        const { children, group, key } = groupProps;
        const groupLabel = getGroupLabel ? getGroupLabel(group) : group;

        return (
          <Fragment key={`group-${key}`}>
            <MUIListSubheader
              data-cy={getComposedDataCy(dataCy, SUBPARTS_MAP.optionGroupLabel, groupLabel)}
              style={{ backgroundColor: theme.palette.background.default }}
            >
              {groupLabel}
            </MUIListSubheader>
            {children}
          </Fragment>
        );
      }}
      renderInput={(inputProps) => {
        const { inputProps: extInputProps } = inputProps;
        const baseStyle: CSSProperties = { width: "100%" };

        if (loading) {
          const forwardedInputProps = {
            ...inputProps,
            inputProps: {
              "data-cy": getComposedDataCy(dataCy, SUBPARTS_MAP.loading),
              style: { ...baseStyle, ...style },
            },
          };

          return (
            <MUISkeleton width="100%">
              <MUITextField
                {...forwardedInputProps}
                margin="normal"
                size={size}
                variant={variant}
                style={{ ...baseStyle }}
              />
            </MUISkeleton>
          );
        }

        const forwardedInputProps = {
          ...inputProps,
          inputProps: { ...extInputProps, "data-cy": dataCy, style: { ...baseStyle, ...style } },
        };

        return (
          <MUITextField
            {...forwardedInputProps}
            label={label}
            margin="normal"
            placeholder={placeholder}
            required={required}
            size={size}
            style={{ width: "100%", ...style }}
            type={type}
            variant={variant}
          />
        );
      }}
      renderOption={(option, { selected }) => {
        if (customOptionRendering) {
          return customOptionRendering(option, selected);
        }

        const optionLabel = getOptionLabel(option);

        return (
          <Fragment key={`option-${optionLabel}`}>
            <Checkbox
              dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.optionCheckbox, optionLabel)}
              disabled
              labelPlacement="end"
              value={selected}
            />
            <Typography dataCy={getComposedDataCy(dataCy, SUBPARTS_MAP.optionLabel, optionLabel)}>
              {optionLabel}
            </Typography>
          </Fragment>
        );
      }}
      value={validateValue(value)}
    />
  );
};

export const SelectWithProps = Select;

export default localized(Select, {
  dataCyShortcut: DATA_CY_SHORTCUT,
  localizableProps: LOCALIZABLE_PROPS,
}) as <T extends any>(props: ISelect<T>) => JSX.Element;
