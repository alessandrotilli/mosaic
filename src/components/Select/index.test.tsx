import renderer from "react-test-renderer";
import { ISelect } from "../../types/Select";
import { getLocalizedTestable } from "../../utils/tests";
import Select, { DATA_CY_DEFAULT, DATA_CY_SHORTCUT } from ".";

const defaultProps: ISelect<string> = {
  multiple: false,
  onChange: jest.fn(),
  options: ["A", "B", "C"],
};

const getSelectTestable = (props?: ISelect<string>, dataCy = DATA_CY_DEFAULT) =>
  getLocalizedTestable(Select, { dataCy, domNode: "input", props: { ...defaultProps } });

// TODO: complete tests

describe("Select test suite:", () => {
  it("default", () => {
    const { element, wrapper } = getSelectTestable();
    // console.log(wrapper.debug());
    expect(wrapper).toHaveLength(1);
    expect("default-props-check").toBeTruthy();

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  xit("dataCy", () => {
    const { element, wrapper } = getSelectTestable({ ...defaultProps, dataCy: "custom" }, "custom");
    expect(wrapper).toHaveLength(1);

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });

  xit("localized", () => {
    const props = { ...defaultProps, localized: true };
    const { element, wrapper } = getSelectTestable({ ...props }, props[DATA_CY_SHORTCUT]);
    // console.log(wrapper.debug());
    expect("localizable-props-check").toBeTruthy();

    const snapshotWrapper = renderer.create(element).toJSON();
    expect(snapshotWrapper).toMatchSnapshot();
  });
});
