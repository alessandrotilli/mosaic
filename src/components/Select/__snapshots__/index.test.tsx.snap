// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select test suite: default 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
  data-cy="select"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <div
    className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-25 MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-controls={null}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        data-cy="select"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={false}
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Clear"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Clear"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-85 MuiOutlinedInput-notchedOutline"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="PrivateNotchedOutline-legend-86"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
